packages/playwright-core/src/server/input.ts

```diff
@@ -44,40 +50,66 @@ export class Keyboard {
 
   async down(key: string) {
     const description = this._keyDescriptionForString(key);
+    if (description.key !== 'Shift') this._deadKeyMappings = undefined;
     const autoRepeat = this._pressedKeys.has(description.code);
     this._pressedKeys.add(description.code);
     if (kModifiers.includes(description.key as types.KeyboardModifier))
       this._pressedModifiers.add(description.key as types.KeyboardModifier);
-    const text = description.text;
-    await this._raw.keydown(this._pressedModifiers, description.code, description.keyCode, description.keyCodeWithoutLocation, description.key, description.location, autoRepeat, text);
+    await this._raw.keydown(this._pressedModifiers, description.code, description.keyCode, description.keyCodeWithoutLocation, description.key, description.location, autoRepeat, description.text);
   }
 
   private _keyDescriptionForString(keyString: string): KeyDescription {
-    let description = usKeyboardLayout.get(keyString);
+    let description = this._keyboardLayout.get(keyString);
     assert(description, `Unknown key: "${keyString}"`);
+    assert(!Array.isArray(description), `Accented key "${keyString}" cannot be typed with a single keypress on "${normalizeLayoutName(this._layoutName)}" keyboard layout.\nUse type() method that will generate multiple key presses.`);
+
     const shift = this._pressedModifiers.has('Shift');
     description = shift && description.shifted ? description.shifted : description;
 
     // if any modifiers besides shift are pressed, no text should be sent
     if (this._pressedModifiers.size > 1 || (!this._pressedModifiers.has('Shift') && this._pressedModifiers.size === 1))
       return { ...description, text: '' };
-    return description;
+
+    if (!this._deadKeyMappings || description.key === 'Shift') return description;
+
+    // handle deadkeys / accented keys
+    const deadKeyText = this._deadKeyMappings.get(description.text);
+
+    // this will be handled properly in the future with IME support.
+    // For now, if there's no dead key mapping for current key, we'll skip it.
+    if (!deadKeyText) return description;
+
+    return { ...description, text: deadKeyText, key: deadKeyText };
   }
 
   async up(key: string) {
     const description = this._keyDescriptionForString(key);
     if (kModifiers.includes(description.key as types.KeyboardModifier))
       this._pressedModifiers.delete(description.key as types.KeyboardModifier);
     this._pressedKeys.delete(description.code);
-    await this._raw.keyup(this._pressedModifiers, description.code, description.keyCode, description.keyCodeWithoutLocation, description.key, description.location);
+    const descKey = description.deadKeyMappings ? 'Dead' : description.key;
+    await this._raw.keyup(this._pressedModifiers, description.code, description.keyCode, description.keyCodeWithoutLocation, descKey, description.location);
+    if (description.key !== 'Shift') this._deadKeyMappings = description.deadKeyMappings;
```

Comment:
Same thing here, I guess.

Reply:
I'm fixing that, and `_deadKeyMappings` will be updated in `keyDown` only. This way it will handle cases like pressing down the deadkey while pressing the key to accent, e.g. `BracketRight+KeyA`. I'm also including a test for that.